common/functions/src/scalars/strings/lower.rs

```diff
@@ -0,0 +1,32 @@
+// Copyright 2021 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use bstr::ByteSlice;
+
+use super::string2string::String2StringFunction;
+use super::string2string::StringOperator;
+
+#[derive(Clone, Default)]
+pub struct Lower;
+
+impl StringOperator for Lower {
+    #[inline]
+    fn apply_with_no_null<'a>(&'a mut self, s: &'a [u8], buffer: &mut [u8]) -> usize {
+        buffer.copy_from_slice(&s.to_lowercase());
```

Comment:
Oh, there is an extra copy.

Maybe we can use  https://docs.rs/bstr/latest/bstr/trait.ByteSlice.html#method.to_lowercase_into

Reply:
After some research, I found:

- `to_lowercase_into` requires a `&mut Vec<u8>` and we can't fit the requirement.

One possible solution is to implement the `to_lowercase_into` by hand. What's your idea?

Reply:
Just use the codes inside  `to_lowercase_into`:

```
 #[inline]
    fn to_lowercase_into(&self, buf: &mut Vec<u8>) {
        // TODO: This is the best we can do given what std exposes I think.
        // If we roll our own case handling, then we might be able to do this
        // a bit faster. We shouldn't roll our own case handling unless we
        // need to, e.g., for doing caseless matching or case folding.

        // TODO(BUG): This doesn't handle any special casing rules.

        buf.reserve(self.as_bytes().len());
        for (s, e, ch) in self.char_indices() {
            if ch == '\u{FFFD}' {
                buf.push_str(&self.as_bytes()[s..e]);
            } else if ch.is_ascii() {
                buf.push_char(ch.to_ascii_lowercase());
            } else {
                for upper in ch.to_lowercase() {
                    buf.push_char(upper);
                }
            }
        }
    }
```